### PR TITLE
[Security] Automatically provide entry points from the custom_authenticator

### LINKED ISSUE #37068: [Security] New LoginFormAuthenticator fails on InsufficientAuthenticationException with missing AuthenticationEntryPoint
**Symfony version(s) affected**: 5.1

**Description**  

I've replaced my guard login authenticator with a version based on the `AbstractLoginFormAuthenticator` and enabled the AuthenticationManager. Everyhing works fine except hitting a page which requires `IS_AUTHENTICATED_REMEMBERED` without actual being logged in.

This triggers an `AccessDeniedException` which gets caught by the firewall ExceptionListener. This creates an `InsufficientAuthenticationException` and tries to  `startAuthentication()` with it. That method then throws an HttpException because there's not AuthenticationEntryPoint defined:

~~~php
        if (null === $this->authenticationEntryPoint) {
            throw new HttpException(Response::HTTP_UNAUTHORIZED, $authException->getMessage(), $authException, [], $authException->getCode());
        }
~~~

The guard system correctly uses the configured entry point to redirect to my login page.

**How to reproduce**

1. Build a login form based on the new authenticator
2. Build a page requiring a authenticated user
3. Hit the page w/o authentication

**Possible Solution**  

Define an entry point for the authenticator?

Sorry if I missed some doc. I've checked the blog post and Wouter's additional article, but couldn't find help.

Kind regards
Matthias

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.1
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #37068 
| License       | MIT
| Doc PR        | -

This PR allows the custom authenticator factory to automatically configure the entry point or fail with an error if more than 1 entry point could be configured. This makes it behave the same as the guard factory.

```yaml
security:
  firewalls:
    main:
      # if only one custom authenticator implements AuthenticationEntryPointInterface,
      # it's automatically configured as the entry point
      custom_authenticator: App\Security\CustomAuthenticator
      custom_authenticators: [App\Security\CustomAuthenticator, App\Security\NoEntryPointAuthenticator]

      # if no custom authenticator implements AuthenticationEntryPointInterface,
      # no entry point is configured automatically
      custom_authenticator: App\Security\NoEntryPointAuthenticator

      # if more than one authenticator implements AuthenticationEntryPointInterface,
      # the entry point must be explicitly set
      custom_authenticators:
        services: [App\Security\CustomAuthenticator, App\Security\AnotherCustomAuthenticator]
        entry_point: App\Security\CustomAuthenticator
```

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/CustomAuthenticatorFactory.php [hunk 1]
```diff
@@ -44,27 +46,55 @@ public function getKey(): string
...
+
+    public function registerEntryPoint(ContainerBuilder $container, string $id, array $config): ?string
+    {
+        if (isset($config['entry_point'])) {
+            return $config['entry_point'];
+        }
+
+        $entryPoints = [];
+        foreach ($config['services'] as $authenticatorId) {
+            if (class_exists($authenticatorId) && is_subclass_of($authenticatorId, AuthenticationEntryPointInterface::class)) {
```
Comment (wouterj) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/CustomAuthenticatorFactory.php:85]: Sadly, we can't do `$container->findDefinition($authenticatorId)->getClass()` here, as the `ContainerBuilder` is scoped for this extension.

Is there some magic trick we can do here to also allow non-class service IDs?

**Comment by wouterj** [comment]
Another suggestion - provided by @althaus:

> What if the listener would add a info/warning to the log before converting it instantly into a HttpEx? Like "Hey there's no entry point set, so I make this an error now. You could set the.config.value to change my behaviour."

**Comment by weaverryan** [comment]
My first impression was: yay! But... when I think further, I realize that eventually, you might have ugly config like this:

```yml
main:
    form_login: ~
    entry_point: form_login
    custom_authenticators:
        services: [App\Security\CustomAuthenticator, App\Security\AnotherCustomAuthenticator]
        entry_point: App\Security\CustomAuthenticator
```

In this case, the `form_login` entrypoint would win... which is fine, but it means that the `entry_point` config under `custom_authenticators` is 100% pointless 🙃 . I really don't like that.

So, I can think of 2 options:

1) On a missing entrypoint, we throw a better exception - basically the suggestion from @althaus. Though... instead of registering a listener to do this, the "authenticator" system could add a "dummy entrypoint" service to the system that would throw this error when called. The idea would be that if (A) you have at least one `custom_authenticator` then the dummy entrypoint is added and (B) if you have *no* other authentication mechanisms, then it would be used (and you would see the error).

2) Or... probably do (1), but also go a step further. If there is ONE `custom_authenticator` (that is an entrypoint), then we automatically register that as an entrypoint. If there are no other authentication mechanisms that provide an entrypoint, it will be used. Yay! As soon as there are TWO `custom_authenticators` (that have an entrypoint), then we register the fake entrypoint described in (1) and tell people to configure the top-level `entry_point` under the firewall. We can even tell them what their options are (i.e. list the 2/3/4 custom authenticator service ids).

So... probably 2 is best. In both cases, I'd like to avoid the `entry_point` that's below `custom_authenticators` as this is really a duplicate of the `entry_point` that's under the firewall (and this was my mistake from the original Guard implementation).

**Comment by wouterj** [comment]
@weaverryan 2 notes and that's why I think improving the log message is the best we can do:

* It's perfectly fine to not have an entry point. This will result in an HTTP Unauthorized response, which is perfect for e.g. APIs
* An authenticator does not have to implement `AuthenticationEntryPointInterface`. So we can't blindly register any authenticator configured using `custom_authenticators`

**Comment by wouterj** [comment]
Closing, see my previous comment on why I no longer believe this PR can ever work. I'll create another one that improves logging.

**Pull request closed without merging**